cmd/syncthing/relays.go

```diff
@@ -0,0 +1,167 @@
+)
+
+func newRelayService(cfg *config.Wrapper, tlsCfg *tls.Config, conns chan<- connection) *relaySvc {
+	svc := &relaySvc{
+		Supervisor: suture.New("relaySvc", suture.Spec{
+			Log: func(log string) {
+				if debugNet {
+					l.Infoln(log)
+				}
+			},
+			FailureBackoff:   5 * time.Minute,
+			FailureDecay:     float64((10 * time.Minute) / time.Second),
+			FailureThreshold: 5,
+		}),
+		cfg:    cfg,
+		tlsCfg: tlsCfg,
+		conns:  conns,
+
+		invitations: make(chan protocol.SessionInvitation),
+	}
+
+	cfg.Subscribe(svc)
+
+	return svc
+}
+
+type relaySvc struct {
+	*suture.Supervisor
+	cfg    *config.Wrapper
+	tlsCfg *tls.Config
+	conns  chan<- connection
+
+	stop    chan struct{}
+	stopped chan struct{}
+
+	tokens      map[string]suture.ServiceToken
+	invitations chan protocol.SessionInvitation
+	running     bool
+}
+
+func (s *relaySvc) Serve() {
+	if debugNet {
+		l.Debugln("Relay service starting")
+	}
+	s.Supervisor.ServeBackground()
+	s.stop = make(chan struct{})
+	s.tokens = make(map[string]suture.ServiceToken)
+
+	rcfg := cfg.Raw()
+	s.CommitConfiguration(rcfg, rcfg)
+
+	for {
+		select {
+		case inv := <-s.invitations:
+			from := syncthingprotocol.DeviceIDFromBytes(inv.From)
+			to := fmt.Sprintf("%s:%d", net.IP(inv.Address).String(), inv.Port)
+			if debugNet {
+				l.Debugln("Received relay invitation from", from, to)
+			}
+			conn, err := client.JoinSession(inv)
+			if err != nil {
+				if debugNet {
+					l.Debugf("Failed to join relay session (%s - %s): %v", from, to, err)
+				}
+				continue
+			}
+
+			setTCPOptions(conn.(*net.TCPConn))
+
+			var tc *tls.Conn
+
+			if inv.ServerSocket {
+				tc = tls.Server(conn, s.tlsCfg)
+			} else {
+				tc = tls.Client(conn, s.tlsCfg)
+			}
+			err = tc.Handshake()
+			if err != nil {
+				l.Infof("TLS handshake (BEP/relay %s - %s): %v", from, to, err)
+				tc.Close()
+				continue
+			}
+			s.conns <- connection{
+				tc, connectionTypeRelayAccept,
+			}
+		case <-s.stop:
+			return
+		}
+	}
+}
+
+func (s *relaySvc) Stop() {
+	if s.stop == nil {
+		return
+	}
+
+	if debugNet {
+		l.Debugln("Relay service stopping")
+	}
+	s.cfg.Unsubscribe(s)
+	s.Supervisor.Stop()
+	close(s.stop)
+	s.stop = nil
+	s.tokens = nil
+}
+
+func (s *relaySvc) VerifyConfiguration(from, to config.Configuration) error {
+	return nil
+}
+
+func (s *relaySvc) CommitConfiguration(from, to config.Configuration) bool {
+	existing := make(map[string]struct{}, len(to.Options.RelayServers))
+	for _, addr := range to.Options.RelayServers {
+		uri, err := url.Parse(addr)
+		if err != nil {
+			if debugNet {
+				l.Debugln("Failed to parse relay address", addr, err)
+			}
+			continue
+		}
+
+		existing[uri.String()] = struct{}{}
+
+		_, ok := s.tokens[uri.String()]
+		if !ok {
+			if debugNet {
+				l.Debugln("Connecting to relay", uri)
+			}
+			c := client.NewProtocolClient(uri, s.tlsCfg.Certificates, s.invitations)
+			s.tokens[uri.String()] = s.Add(&c)
+		}
+	}
+
+	for uri, token := range s.tokens {
```

Comment:
No, I mean like in gui.go for example. Do the prep work in CommitConfiguration, then tell Serve to exit. Serve does whatever cleanup necessary on the way out - as it would in a panic or clean shutdown. The supervisor restarts it. Everyone is happy.